### PR TITLE
Initial addition of 9p code to Podman

### PR DESCRIPTION
**Pull request by mheon**
This includes two new hidden commands: a 9p server, `podman system server9`, and a 9p client, `podman system client9` with `server9` currently only configured to run on Windows and serve 9p via HyperV vsock, and `client9` only configured to run on Linux. The server is run by `podman machine start` and has the same lifespan as gvproxy (waits for the gvproxy PID to die before shutting down). The client is run inside the VM, also by `podman machine start`, and mounts uses kernel 9p mount code to complete the mount. It's unfortunately not possible to use mount directly without the wrapper; we need to set up the vsock and pass it to mount as an FD.

In theory this can be generalized so that the server can run anywhere and over almost any transport, but I haven't done this here as I don't think we have a usecase other than HyperV right now.

#### Does this PR introduce a user-facing change?

```release-note
Podman machine VMs created using HyperV can now share directories with the host
```

### CODE REVIEW

**Comment by mheon** [comment]
WIP as I am still trying to actually get things running reliably on Windows.

Test situation is suboptimal, as we need a Podman with my changes inside the `podman machine` VM we test with. Could potentially use a custom image?

**Review by Luap99 — COMMENTED** [review]
Just some small comments for the cli.
Also can we move these commands under `podman machine`? I feel like this makes more sense logicically and in case we ever decide to split out machine into a separate repo.
The core podman experience has no need for these commands.

**Comment by mheon** [comment]
Moving commands to machine SGTM, don't know why I didn't think of that before.

**Comment by mheon** [comment]
`bin/podman grew by 81672 bytes; max allowed is 51200.`
I'm going to bloat-approved given this isn't that much, but worth noting.

**Comment by mheon** [comment]
This seems to work, but I'm going to hold off on removing WIP until I can retest tomorrow to verify.

**Comment by mheon** [comment]
Dropping WIP

File: cmd/podman/machine/9client.go
Comment (giuseppe) [inline cmd/podman/machine/9client.go:120]: should this condition be inverted?

Comment (mheon) [reply cmd/podman/machine/9client.go]: Good catch, done

File: cmd/podman/machine/9client.go
Comment (gbraad) [inline cmd/podman/machine/9client.go:25]: `client9p`

File: cmd/podman/machine/9client.go
Comment (mheon) [reply cmd/podman/machine/9client.go]: Done

**Comment by mheon** [comment]
This is ready for review + merge. Once that's done, we can move the backend code out into libhvee (or somewhere else, given it's not really Windows-specific) for reuse by CRC.

**Comment by mheon** [comment]
@baude @ashley-cui @Luap99 PTAL

**Comment by TomSweeneyRedHat** [comment]
Man page changes?
Quick breeze review looks good, but definitely want other eyeballs.

File: pkg/machine/hyperv/machine.go
Comment (Luap99) [inline pkg/machine/hyperv/machine.go:786]: Can I just say that I hate having applications write directly into the home dir. Shouldn't this be but into the machine dir itself? At the very least so it gets removed when you remove the machine but also so it doesn't conflict if we can ever start more than one machine at once.

Second because you duplicate the same code twice can you split this into a function and please add a log line that says logging output from X to file Y. I have no idea how much these programs will write but I can imagine someone running out of disk space if they are not aware of this.

**Comment by Luap99** [comment]
Can you squash the commits so that they make logically sense once merged. Bug fixes for things added in a prior commit should always just be squashed into it and not added as extra commit. Otherwise you may break an easy git bisect.

Also I see no reason for a extra commit to move the commands to the machine subcommand, that should also be just squashed.
`podman system server9` should not become part of the git log as it never will exist realistically once merged.

File: cmd/podman/machine/client9p.go
Comment (Luap99) [inline cmd/podman/machine/client9p.go:127]: This is logging the error and passing it back via channel, thus it will be logged twice.
Instead of logging this just return this as fmt.Errorf()

File: cmd/podman/machine/server9p.go
Comment (Luap99) [inline cmd/podman/machine/server9p.go:96]: I am not quite sure why this is needed? There seems to be a WaitForExit() call in windows which unlike linux should also work if the process you are waiting for is not a child. Are you something like https://stackoverflow.com/questions/37664015/os-process-wait-after-os-findprocesspid-works-on-windows-not-on-linux does not work and we really need such a big dep?

Alternative why can't we just use the same logic as for gvproxy stop and let the machine commands kill this process and let it sleep forever here instead of this busy loop.

File: cmd/podman/machine/client9p.go
Comment (Luap99) [inline cmd/podman/machine/client9p.go:118]: Wouldn't it be simpler to just use cmd.CombinedOutput() instead? I doubt anybody needs to care about the stdout of mount

**Review by Luap99 — COMMENTED** [review]
Not a full review just some drive by comments

File: cmd/podman/machine/client9p.go
Comment (mheon) [reply cmd/podman/machine/client9p.go:127]: It's not logging the error, it's logging stderr from the command - actual error is the exit code of the command.

File: cmd/podman/machine/server9p.go
Comment (mheon) [reply cmd/podman/machine/server9p.go:96]: The WaitForExit call is not accessible via Golang's standard library without DLL loading shenanigans and then calling untyped functions (well, the function itself is, but the ability to get the handle that it requires to work is not) - so this is *a lot* cleaner, even if it's technically gross.

I don't have objections to using a PID file and terminating on machine stop, though. @baude You have strong feelings on this?

File: cmd/podman/machine/client9p.go
Comment (Luap99) [reply cmd/podman/machine/client9p.go:127]: Yes which means the real error returned by the podman command is a useless exit status X and the useful message is just logged on stderr above. And both lines log the same error condition!

IMO the right things is take that message and append that to a actual error with fmt.Errorf() and then return the error.

File: cmd/podman/machine/server9p.go
Comment (mheon) [reply cmd/podman/machine/server9p.go:96]: Further, I'm strongly in favor of keeping the dep even if we drop this code. The things we were doing with Go's native `os.Process` code to do things like check for process exit just did not work - I solved 2 separate bugs elsewhere in the code by replacing our custom-rolled Windows/Linux process code with this library.

Comment (Luap99) [reply cmd/podman/machine/server9p.go:96]: I you think it is needed sure, not like I tested it on windows. I just cannot believe that the go std lib wouldn't suffice for that use case but if it is like that then sure.

Comment (mheon) [reply cmd/podman/machine/server9p.go:96]: From what I can see, the standard library is fine on basically any POSIX-compliant system, but it starts to fall apart on Windows - things are sufficiently different that the abstractions can't handle it.

**Comment by mheon** [comment]
Squashed down, comments addressed

**Comment by mheon** [comment]
@baude PTAL

File: cmd/podman/machine/client9p.go
Comment (Luap99) [inline cmd/podman/machine/client9p.go:110]: This is not right CombinedOutput() already calls Run() for you so you should not call Run() before that.

**Review by Luap99 — CHANGES_REQUESTED** [review]

File: cmd/podman/machine/server9p.go
Comment (gbraad) [reply cmd/podman/machine/server9p.go:96]: I agree here, the stdlib on Windows does not provide the same functionality as you have on Posix-compliant systems. I think this is expectable for now, as we might spin this code out at a later time.

File: pkg/machine/hyperv/machine.go
Comment (gbraad) [reply pkg/machine/hyperv/machine.go]: What gets logged here? For CRC we have a similar log file and this can be a lot. We get a NTP (UDP) timeout a few times once in a while. This causes our log to balloon to several megabytes for a day.

**Comment by TomSweeneyRedHat** [comment]
@cevich the scan for secret leaks and changes test is failing with:

" Error: unrecognized namespace mode keep-id:uid=1000,gid=1000 passed
Error: Process completed with exit code 125."

I'm not sure where that's being thrown from, do you?

**Comment by cevich** [comment]
> I'm not sure where that's being thrown from, do you?

Yes I'm on top of it.  A fix PR was just merged.  A re-run might fix it (I'll try pushing the button), otherwise a force-push should.

**Comment by cevich** [comment]
:frowning_face: no go.  You can ignore the result, or force-push and it will clear.

**Comment by mheon** [comment]
Should merge without it

**Pull request merged**
